pkg/templates/nstemplatetiers/nstemplatetier_generator.go

```diff
@@ -164,43 +186,47 @@ func (g nstemplatetierGenerator) newNSTemplateTiers(namespace string) (map[strin
 //       template: >
 //         <yaml-ns-template>
 // ------
-func (g nstemplatetierGenerator) newNSTemplateTier(tier, namespace string) (*toolchainv1alpha1.NSTemplateTier, error) {
-	revisions, ok := g.revisions[tier]
-	if !ok {
-		return nil, errors.Errorf("tier '%s' does not exist", tier)
+func newNSTemplateTier(decoder runtime.Decoder, namespace, tier string, tmpls templates) (*toolchainv1alpha1.NSTemplateTier, error) {
+	// retrieve the namespace types and order them, so we can compare
+	// with the expected templates during the tests
+	namespaceKinds := make([]string, 0, len(tmpls.namespaceTemplates))
+	for kind := range tmpls.namespaceTemplates {
+		namespaceKinds = append(namespaceKinds, kind)
 	}
-	obj := &toolchainv1alpha1.NSTemplateTier{
+	sort.Strings(namespaceKinds)
+	result := &toolchainv1alpha1.NSTemplateTier{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      tier,
 			Namespace: namespace,
 		},
 		Spec: toolchainv1alpha1.NSTemplateTierSpec{},
 	}
-	// retrieve the namespace types and order them, so we can order them, and compare
-	// with the expected templates during the tests
-	nsTypes := make([]string, 0, len(revisions))
-	for nsType := range revisions {
-		nsTypes = append(nsTypes, nsType)
-	}
-	sort.Strings(nsTypes)
-	for _, nsType := range nsTypes {
-		// get the content of the `-<t>.yaml` file
-		tmpl, err := g.asset(fmt.Sprintf("%s-%s.yaml", tier, nsType))
+	for _, kind := range namespaceKinds {
+		// convert the content into a templatev1.Template
+		tmplObj := &templatev1.Template{}
+		_, _, err := decoder.Decode(tmpls.namespaceTemplates[kind].content, nil, tmplObj)
 		if err != nil {
 			return nil, errors.Wrapf(err, "unable to generate '%s' NSTemplateTier manifest", tier)
 		}
+		// add it to the NSTemplateTier obj
+		result.Spec.Namespaces = append(result.Spec.Namespaces, toolchainv1alpha1.NSTemplateTierNamespace{
+			Type:     kind,
+			Revision: tmpls.namespaceTemplates[kind].revision,
+			Template: *tmplObj,
+		})
+	}
+	// also, add the cluster resource template+revision if it exists
+	if tmpls.clusterTemplate.content != nil {
 		// convert the content into a templatev1.Template
 		tmplObj := &templatev1.Template{}
-		_, _, err = g.decoder.Decode(tmpl, nil, tmplObj)
+		_, _, err := decoder.Decode(tmpls.clusterTemplate.content, nil, tmplObj)
 		if err != nil {
 			return nil, errors.Wrapf(err, "unable to generate '%s' NSTemplateTier manifest", tier)
 		}
-		// add it to the NSTemplateTier obj
-		obj.Spec.Namespaces = append(obj.Spec.Namespaces, toolchainv1alpha1.NSTemplateTierNamespace{
-			Type:     nsType,
-			Revision: revisions[nsType],
+		result.Spec.ClusterResources = toolchainv1alpha1.NSTemplateTierClusterResources{
```

Comment:
To be honest, I still don't understand why we have to use a pointer (I know it works), whereas in 
```
// NSTemplateSetNamespace the namespace definition in an NSTemplateSet resource
type NSTemplateSetNamespace struct {
	// The type of the namespace. For example: ide|cicd|stage|default
	Type string `json:"type"`

	// The revision of the corresponding template
	Revision string `json:"revision"`

	// Optional field. Used to specify a custom template
	Template string `json:"template,omitempty"`
}
```
the `Template` field is correctly omitted when empty and it works. In other words, I would like to make sure that the pointer is really the way we have to proceed here. We don't use pointers anywhere else in our API types, so I'm a bit surprised we need that now.

Reply:
ok, so as mentioned in https://github.com/kubernetes/community/blob/6a41010589b0ba425fbf8b84872538e586619d45/contributors/devel/sig-architecture/api-conventions.md:
> Using a pointer allows distinguishing unset from the zero value for that type. There are some cases where, in principle, a pointer is not needed for an optional field since the zero value is forbidden, and thus implies unset. There are examples of this in the codebase. However:
> - it can be difficult for implementors to anticipate all cases where an empty value might need to be distinguished from a zero value
> - structs are not omitted from encoder output even where omitempty is specified, which is messy;
> - having a pointer consistently imply optional is clearer for users of the Go language client, and any other clients that use corresponding types
>
> Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.

so I'll incorporate your change in https://github.com/codeready-toolchain/host-operator/commit/f93bbe11a15087b1783692393d5588d45f7f38e4 ;)

Reply:
done in fef2870

Reply:
To addition to that quote the problem here is that `omitempty` works only with specific types:

> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string.

See https://golang.org/pkg/encoding/json/
So it doesn't work with empty arbitrary structs like `Template`. And it's not just theoretical. If you switch to an empty template vs. a pointer and remove ClusterResources yaml from the team template then our e2e tests will fail with the following error:
```
{"level":"error","ts":1583774144.763569,"logger":"cmd","msg":"","error":"unable to create the NSTemplateTiers 'basic' in namespace 'codeready-toolchain-host-09165437': NSTemplateTier.toolchain.dev.openshift.com \"basic\" is invalid: spec.clusterResources.template.objects: Invalid value: \"null\": spec.clusterResources.template.objects in body must be of type array: \"null\"","errorVerbose":"NSTemplateTier.toolchain.dev.openshift.com \"basic\" is invalid: spec.clusterResources.template.objects: Invalid value: \"null\": spec.clusterResources.template.objects in body must be of type array: \"null\"\nunable to create the NSTemplateTiers 'basic' in namespace 'codeready-toolchain-host-09165437'\ngithub.com/codeready-toolchain/host-operator/pkg/templates/nstemplatetiers.CreateOrUpdateResources\n\t/tmp/codeready-toolchain/host-operator/pkg/templates/nstemplatetiers/nstemplatetier_generator.go:40\nmain.main.func1\n\t/tmp/codeready-toolchain/host-operator/cmd/manager/main.go:193\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nmain.main.func1\n\t/tmp/codeready-toolchain/host-operator/cmd/manager/main.go:194"} 
``` 
It's because the ClusterResources fields is not omitted but it contains an empty Template with empty objects array which is forbidden, so the tier doesn't load and the operator doesn't start.

Reply:
yes, that's the error I got (before using a pointer), indeed.